cunumeric/module.py

```diff
@@ -2677,11 +2819,31 @@ def einsum(expr, *operands, out=None, optimize=False):
         expr, *operands, einsum_call=True, optimize=optimize
     )
     for (indices, _, sub_expr, _, _) in contractions:
-        sub_opers = [operands.pop(i) for i in indices]
-        if len(operands) == 0:  # last iteration
-            sub_result = _contract(sub_expr, *sub_opers, out=out)
+        assert len(indices) == 1 or len(indices) == 2
+        a = operands.pop(indices[0])
+        b = operands.pop(indices[1]) if len(indices) == 2 else None
+        if b is None:
+            m = re.match(r"([a-zA-Z]*)->([a-zA-Z]*)", sub_expr)
```

Comment:
Do we know `re.match` internally caches the internal representation for this match string? If we're gonna use this match string repeatedly, it wouldn't be a terrible idea to cache the matcher ourselves with `re.compile`.

Reply:
Yes, python internally compiles and caches the regexes passed to `re` functions. Quoting from https://docs.python.org/3/library/re.html#re.compile:

```
Note The compiled versions of the most recent patterns passed to [re.compile()](https://docs.python.org/3/library/re.html#re.compile) and the module-level matching functions are cached, so programs that use only a few regular expressions at a time needn’t worry about compiling regular expressions.
```

That's why I don't generally bother compiling regexes manually, but I suppose compiling manually would save us a dictionary lookup.

Reply:
I'll just leave it as it is for now, given that the gains from explicit compilation are likely minimal.